### PR TITLE
show processed token in greeting dropdown

### PR DESCRIPTION
**Pull request by sunilpawar**
Overview
----------------------------------------
For Greeting we show Dear with some contact token as configured in greeting back-end section. Some time to generate custom greeting, we need to write smarty code with conditional logic capture smarty function. On Contact Summary page, same smarty code shown in greeting drop down list.

So instead of showing smarty code we can show processed greeting in drop drown list and user don't have check logic written for each greeting block. Description will shown in bracket if present in-front of each processed greeting.

Before
----------------------------------------
Tokens were displayed with conditional logic if present ...
Screenshot 2021-09-27 at 7 16 43 PM

After
----------------------------------------
Token get Processed and processed token output get shown.

Screenshot 2021-09-27 at 7 16 20 PM

Back-end configuration of Addressee Token:
Screenshot 2021-09-27 at 9 00 46 PM

Comments
----------------------------------------
I did submitted Same PR few year back with similar changes but failed in user testing:
https://github.com/civicrm/civicrm-core/pull/10480

Relevant Ticket link on JIRA  : https://issues.civicrm.org/jira/browse/CRM-20695

### CODE REVIEW

File: CRM/Contact/Form/Edit/CommunicationPreferences.php [hunk 1]
```diff
@@ -82,8 +82,33 @@ public static function buildQuickForm(&$form) {
...
+      if ($form->_contactId) {
+        $contact = new CRM_Contact_BAO_Contact();
+        $contact->id = $form->_contactId;
+        $contact->find(TRUE);
+        $contactDetails = [];
+        CRM_Core_DAO::storeValues($contact, $contactDetails);
+        $contactDetails = [[$contact->id => $contactDetails]];
+        foreach ($greetingTokens as $key => &$emailGreetingString) {
+          if ($emailGreetingString) {
+            CRM_Contact_BAO_Contact_Utils::processGreetingTemplate($emailGreetingString,
```
Comment (eileenmcnaughton) [inline CRM/Contact/Form/Edit/CommunicationPreferences.php:101]: The bad news is that I think that function is one of the ones we are removing this release. If you leave it a couple of weeks we should be clearer with guidance on that - we would need a unit test though

Comment (sunilpawar) [reply CRM/Contact/Form/Edit/CommunicationPreferences.php:101]: ok, however there is nothing mentioned in function doc about `@deprecated` for function `processGreetingTemplate` ?

Comment (eileenmcnaughton) [reply CRM/Contact/Form/Edit/CommunicationPreferences.php:101]: @sunilpawar no - defintely hoping to have that sorted by the time we cut 5.43 - just a couple more things to iron out

Comment (totten) [reply CRM/Contact/Form/Edit/CommunicationPreferences.php:101]: I haven't tested, but I think this revision would switch it to the newer mechanism:

```diff
diff --git a/CRM/Contact/Form/Edit/CommunicationPreferences.php b/CRM/Contact/Form/Edit/CommunicationPreferences.php
index 7e232de0af..c9fad4bca9 100644
--- a/CRM/Contact/Form/Edit/CommunicationPreferences.php
+++ b/CRM/Contact/Form/Edit/CommunicationPreferences.php
@@ -90,17 +90,12 @@ class CRM_Contact_Form_Edit_CommunicationPreferences {
       // Get Description of each greeting.
       $greetingTokensDescription = CRM_Core_PseudoConstant::greeting($filter, 'description');
       if ($form->_contactId) {
-        $contact = new CRM_Contact_BAO_Contact();
-        $contact->id = $form->_contactId;
-        $contact->find(TRUE);
-        $contactDetails = [];
-        CRM_Core_DAO::storeValues($contact, $contactDetails);
-        $contactDetails = [[$contact->id => $contactDetails]];
+        $renderedGreetingTokens = CRM_Core_TokenSmarty::render($greetingTokens, [
+          'contactId' => $form->_contactId,
+        ]);
         foreach ($greetingTokens as $key => &$emailGreetingString) {
           if ($emailGreetingString) {
-            CRM_Contact_BAO_Contact_Utils::processGreetingTemplate($emailGreetingString,
-              $contactDetails, $form->_contactId, 'CRM_Contact_BAO_Contact'
-            );
+            $emailGreetingString = $renderedGreetingTokens[$key];
             $emailGreetingString = CRM_Core_DAO::escapeString(CRM_Utils_String::stripSpaces($emailGreetingString));
             if (!empty($greetingTokensDescription[$key])) {
               // Append description to processed greeting.
```

But that does leave another question (sort of pre-existing/tangential to this PR) -- `CRM_Contact_BAO_Contact_Utils::processGreetingTemplate` is used by a few more places  - do we switch all of these at once, switch gradually, or do some refactoring to nominally preserve the function. (AFAICS, there's nothing in `universe` which calls this function, aside from `civicrm-core`.)

Comment (sunilpawar) [reply CRM/Contact/Form/Edit/CommunicationPreferences.php:101]: @totten your changes work perfectly. I will update the PR with new changes...

**Pull request merged**